### PR TITLE
Touch interaction: allow users to disable touch support completely

### LINKED ISSUE #9682: How to prevent NVDA to overwrite touchscreen configurations on Windows
### Steps to reproduce:
When NVDA is on, it overwrites a set of predefined touch and gestures configurations for touchscreens, so what was before a single tap action becomes a fast double tap one. I need to prevent NVDA from overwriting some default settings on Windows.

I already tried removing "tap (Touch screen)" from "Input Gestures" under "Preferences" option, but it only stopped dictating, I still have to double tap to perform an action.

### System configuration
I'm using windows 10

#### NVDA version:
2019.1.1

#### Windows version:
Windows 10

### PR DESCRIPTION
**Pull request by josephsl**
### Link to issue number:
Fixes #9682 

### Summary of the issue:
Allow users to turn off touch support completely.

### Description of how this pull request fixes the issue:
Added a checkbox (and a corresponding flag in NVDA configuration database) that disables touch support completely. If disabled, touchscreen will behave as though NVDA is not running. Also, added support for touch being disabled on per-profile basis i.e. touch can be disabled if an app comes with its own touch interaction model.

### Testing performed:
Tested via try builds on a laptop with touch capabilities.

### Known issues with pull request:
None so far.

### Change log entry:
Changes (although this can go as a new feature):

You can now enable or disable NVDA's touchscreen support. This is done by checking a new checkbox in Touch Interaction panel of NVDA settings. (#9682)

Thanks.

### CODE REVIEW

**Comment by josephsl** [comment]
Hi,

This will be considered "draft" while NVDA 2019.3 is under development; once stable version is released, draft status will be removed. IN the meantime, comments (including design comments) are appreciated.

Thanks.

File: source/touchHandler.py [hunk 1]
```diff
@@ -294,30 +293,56 @@ def notifyInteraction(self, obj):
 
 handler=None
 
-def touchSupported():
+
+def touchSupported(debugLog=False):
```
Comment (leonardder) [inline source/touchHandler.py]:
```suggestion
def touchSupported(debugLog: bool = False):
```

File: source/touchHandler.py [hunk 2]
```diff
@@ -294,30 +293,56 @@ def notifyInteraction(self, obj):
 
 handler=None
 
-def touchSupported():
+
+def touchSupported(debugLog=False):
 	"""Returns if the system and current NVDA session supports touchscreen interaction.
```
Comment (leonardder) [inline source/touchHandler.py]:
```suggestion
	"""Returns if the system and current NVDA session supports touchscreen interaction.
	@param debugLog: Whether to log additional details about touch support to the NVDA log.
```

File: source/touchHandler.py [hunk 3]
```diff
@@ -294,30 +293,56 @@ def notifyInteraction(self, obj):
...
+	if not handler and enable:
+		handler = TouchHandler()
+		log.debug("Touch support enabled.")
+	elif handler and not enable:
+		handler.terminate()
+		handler = None
+		log.debug("Touch support disabled.")
+
+
+def handlePostConfigProfileSwitch():
```
Comment (leonardder) [inline source/touchHandler.py]: Do we execute a config profile switch when saving config from the gui? I recall we don't.

File: source/touchHandler.py [hunk 4]
```diff
@@ -294,30 +293,56 @@ def notifyInteraction(self, obj):
...
 	if maxTouches<=0:
-		log.debugWarning("No touch devices found")
+		if debugLog:
+			log.debugWarning("No touch devices found")
 		return False
 	return True
 
-def initialize():
+
+def setTouchSupport(enable):
```
Comment (leonardder) [inline source/touchHandler.py]:
```suggestion
def setTouchSupport(enable: bool):
```

**Review by LeonarddeR — CHANGES_REQUESTED** [review]
Could you please change the touch panel to guiHelper while at it?

**Comment by josephsl** [comment]
Hi, I don’t think we execute profile switch handlers when settings are saved – I borrowed this logic from Speech category/audio ducking. Thanks.

File: user_docs/en/userGuide.t2t [hunk 5]
```diff
@@ -1494,6 +1495,11 @@ If you check this option and you have the "Enable mouse tracking" option enabled
 This settings category, only available on computers running Windows 8 and later with touch capabilities, allows you to configure how NVDA interacts with touchscreens.
 This category contains the following options:
 
+==== Enable touch interaction support ====[TouchSupportEnable]
+This checkbox enables NVDA's touch interaction support.
+If enabled, you can use your fingers to navigate and interact with items on screen using a touchscreen device.
+If disabled, touchscreen support from NVDA will be disabled.
```
Comment (leonardder) [inline user_docs/en/userGuide.t2t]: What does this mean?

```suggestion
If disabled, touchscreen support from NVDA will be disabled, making the touch screen behave as normal.
```

**Comment by josephsl** [comment]
Hi, I had difficulty when writing that sentence, and the suggestion above confirms my thought that we need to clarify a bit. Thanks for pointing this out.

**Comment by josephsl** [comment]
Hi,

Now that 2019.3 is under translatable string freeze, and with 2020.1 changelog showing, I'd like to get this PR reviewed again.

Thanks.

**Comment by josephsl** [comment]
Hi,

Suspended until further notice in light of #11006. 

Thanks.

**Comment by feerrenrut** [comment]
> Suspended until further notice in light of #11006.

Please continue to work on this PR. #11006 requests no new PR's. This PR already exists and is thus in the backlog of PR's that we want to address before accepting new ones. We are trying to stop PR's from going stale, and then becoming hard to merge. 

Forgive me for replying to you multiple times, I want to avoid the community being mislead.

**Comment by josephsl** [comment]
Hi, I’ll take a look at this PR as early as Thursday. Thanks.

**Comment by dingpengyu** [comment]
> Hi,
> 
> Suspended until further notice in light of #11006.
> 
> Thanks.

hello josephsl
You currently have the following pr not merged yet:
#10990 
#10922 
#10378 
#10349 
#9429 
#9119 
#8644 
#8457 
and
#6930 
Nine in total pull request.
In addition to the current COVID-19 epidemic, there is ample time.

**Comment by josephsl** [comment]
Hi, I’ll evaluate each of those and comment on status of these soon – rest assured that I’m thinking about them. Thanks for the reminder.

**Comment by JulienCochuyt** [comment]
@josephsl, thanks for taking care of this. The lack of this feature has been annoying me quite a lot lately.

My few grains of salt on this topic...
- Would you please consider allowing to bind a gesture on toggling touch screen support on and off?
- Shouldn't we consider to temporarily disable touch screen support when pressing `NVDA+F2`? If not, any suitable mean of letting a touch gesture through would be really handy.
- Does UIA or IA2 expose the fact that a given control should have touch gestures not intercepted by AT? (such as found on touch musical instruments in IOS) If so, we should most likely plan on honoring it.
- Should we consider shipping a default profile for Windows Touch Magnifier window to disable touch support when interacting with it?

**Comment by josephsl** [comment]
Hi,

Suspending touch support for passthrough command: not easy. In order to do so, NVDA must be able to detect that touch interaction has occurred after it is suspended. This isn't quite possible because once an assistive technology registers for touch interaction, it won't let go unless it voluntarily gives up, and manual work is involved to bring touch support back online. Also, if multiple assistive technologies wants to work with touch interaction, only one will get a chance to do it (when it starts, Narrator presents a dialog about this when NVDA is running on a touchscreen computer).

Command to toggle touch support: although it makes sense, it isn't consistent. One can assign a keyboard or a braille command to toggle touch interaction. One can turn off touch interaction using touch gestures, but you can't bring it back with touch gestures as NVDA doesn't know if a gesture has occurred (same problem as above).

UIA/IA2 and touch support: probably not unless I read the documentation wrong.

As for Windows Magnifier window: one way to enhance this in the future is letting app modules or overlay classes define if touch interaction should be turned off. This is already possible with this PR through app-specific profiles, which users must specify manually.

Thanks.

**Comment by JulienCochuyt** [comment]
@josephsl wrote:
> Suspending touch support for passthrough command: not easy. In order to do so, NVDA must be able to detect that touch interaction has occurred after it is suspended. This isn't quite possible because once an assistive technology registers for touch interaction, it won't let go unless it voluntarily gives up, and manual work is involved to bring touch support back online. 

Letting other ATs aside, do you mean there is no such mechanism as sending a touch gesture we received back to the control that would have received it if NVDA did not register?
If so, I think my behavior of choice would be to (at least optionally) suspend touch handling until the next keystroke, to allow for keyboard+touch usages.

> Command to toggle touch support: although it makes sense, it isn't consistent. One can assign a keyboard or a braille command to toggle touch interaction. One can turn off touch interaction using touch gestures, but you can't bring it back with touch gestures as NVDA doesn't know if a gesture has occurred (same problem as above).

Idem, I feel that inconsistency here is better than inability, especially if we cannot achieve a "next touch gesture pass through" command.

> As for Windows Magnifier window: one way to enhance this in the future is letting app modules or overlay classes define if touch interaction should be turned off. This is already possible with this PR through app-specific profiles, which users must specify manually.

Glad to read this. We for sure can take the time once this PR is merged to experiment with such a profile before deciding if it's worth being promoted as a default.

**Comment by josephsl** [comment]
Hi,

Once an AT unregisters itself from getting touch gestures, it isn't possible to listen to touch gestures. This is because an invisible iwndow is used by the AT to listen to touch interaction signals, and suspending touch means destroying that window.

As for toggle command: I imagine a scenario where people would prefer to use touchscreen alone rather than using a keyboard along with the device (Surface Pro without type cover, for example).

Thanks.

**Comment by JulienCochuyt** [comment]
> Once an AT unregisters itself from getting touch gestures, it isn't possible to listen to touch gestures. This is because an invisible iwndow is used by the AT to listen to touch interaction signals, and suspending touch means destroying that window.

Sure, but again, do you mean we cannot send back what we don't want to keep (as keyboard handler does)?

> As for toggle command: I imagine a scenario where people would prefer to use touchscreen alone rather than using a keyboard along with the device (Surface Pro without type cover, for example).

While this scenario makes sense, it's no reason IMHO to disallow altogether binding a keyboard gesture to do the toggling.

**Comment by josephsl** [comment]
Hi, no, we can’t do that with touchscreen gestures at this time, as the window itself isn’t going to be used. In short, Windows will forget that an AT is active. As for toggle command bindings, it might be possible to do that at least for keyboards – it is something we should think about for a future iteration of this PR. Thanks.

**Comment by michaelDCurran** [comment]
@JulienCochuyt  You bring up all good points, but these should be all handled in one or more separate PRs. Not only are they a little complex and will be somewhat technically unrelated to how this is currently implemented. By itself this current PR still has a clear use case. There are certainly low vision users who use NVDA to read text along with a magnifier, but have enough vision to interact with their touch screen independently and do not want NVDA ever intercepting touch. So @josephsl   assuming this passes review, I want to take this as soon as we can.

**Review by michaelDCurran — APPROVED** [review]

**Comment by michaelDCurran** [comment]
@josephsl  just confirming that you have tested config profile switches where one profile has touch enabled and one does not? I.e. touch successfully gets re-enabled after being disabled by a previous profile?

**Comment by josephsl** [comment]
Hi, glad to contribute another feature from Enhanced Touch Gestures to NVDA Core – thanks.

**Comment by josephsl** [comment]
Hi, yes, tested via Enhanced Touch Gestures add-on, which has identical construction. Thanks.

**Comment by josephsl** [comment]
Hi, forgot to say that profile switching does work. Thanks.

**Comment by michaelDCurran** [comment]
@leonardder  Are there any pending changes you are waiting for on this pr?

File: source/touchHandler.py [hunk 6]
```diff
@@ -294,30 +293,57 @@ def notifyInteraction(self, obj):
...
+	@param debugLog: Whether to log additional details about touch support to the NVDA log.
 	"""
 	if not config.isInstalledCopy() and not config.isAppX:
-		log.debugWarning("Touch only supported on installed copies")
+		if debugLog:
+			log.debugWarning("Touch only supported on installed copies")
 		return False
-	if (winVersion.winVersion.major*10+winVersion.winVersion.minor)<62:
-		log.debugWarning("Touch only supported on Windows 8 and higher")
+	if winVersion.winVersion.platform_version < (6, 2, 9200):
```
Comment (leonardder) [inline source/touchHandler.py:305]: Is it really necessary to take the platform into account here?

```suggestion
	if (winVersion.winVersion.major, winVersion.winVersion.minor) < (6, 2):
```

**Review by LeonarddeR — APPROVED** [review]
If a sighted person can confirm that the gui looks good (though I don't have reason to think it wouldn't), I'm ok with this. Just one small thing.

**Comment by josephsl** [comment]
Hi, the platform attribute gives us all the info we need in one function call, and it is also used for app module version lookup. Thanks.

**Pull request merged**